components/tidb_query_expr/src/impl_time.rs

```diff
@@ -156,6 +157,117 @@ pub fn to_seconds(ctx: &mut EvalContext, t: &DateTime) -> Result<Option<Int>> {
     Ok(Some(t.second_number()))
 }
 
+// is_duration returns a boolean indicating whether the str matches the format of duration.
+// See https://dev.mysql.com/doc/refman/5.7/en/time.html
+#[inline]
+fn is_duration(s: &str) -> bool {
+    let re =
+        Regex::new(r"^\s*[-]?(((\d{1,2}\s+)?0*\d{0,3}(:0*\d{1,2}){0,2})|(\d{1,7}))?(\.\d*)?\s*$")
+            .unwrap();
+    re.is_match(s)
+}
+#[rpn_fn(writer, capture = [ctx])]
+#[inline]
+pub fn add_string_and_duration(
+    ctx: &mut EvalContext,
+    arg0: BytesRef,
+    arg1: &Duration,
+    writer: BytesWriter,
+) -> Result<BytesGuard> {
+    match is_duration(std::str::from_utf8(arg0).unwrap()) {
```

Comment:
According to my benchmark:

`is_duration` is much slower than `Duration::parse`. (If I didn't make some mistakes here)

Maybe we can reduce the `is_duration` check here, use `Duration::parse` directly, if failed, use `DateTime::parse_datetime` instead, if failed again, just return `None`.

---
Result

```rust
test codec::mysql::duration::benches::bench_parse                                         ... bench:       1,742 ns/iter (+/- 76)
test codec::mysql::duration::benches::bench_is_durtion                                    ... bench:   4,452,029 ns/iter (+/- 1,344,206)
```

Code

```rust
#[cfg(test)]
mod benches {
    use super::*;
    use crate::codec::mysql::MAX_FSP;

    #[bench]
    fn bench_parse(b: &mut test::Bencher) {
        let cases = vec![
            ("12:34:56.1234", 0),
            ("12:34:56.789", 1),
            ("10:20:30.189", 2),
            ("2 27:54:32.828", 3),
            ("2 33:44:55.666777", 4),
            ("112233.445566", 5),
            ("1 23:12.1234567", 6),
        ];

        b.iter(|| {
            let cases = test::black_box(&cases);
            for &(s, fsp) in cases {
                let _ = test::black_box(
                    Duration::parse(&mut EvalContext::default(), s.as_bytes(), fsp).unwrap(),
                );
            }
        })
    }
    use regex::Regex;
    fn is_duration(s: &str) -> bool {
        let re =
            Regex::new(r"^\s*[-]?(((\d{1,2}\s+)?0*\d{0,3}(:0*\d{1,2}){0,2})|(\d{1,7}))?(\.\d*)?\s*$")
                .unwrap();
        re.is_match(s)
    }

    #[bench]
    fn bench_is_durtion(b: &mut test::Bencher) {
        let cases = vec![
            ("12:34:56.1234", 0),
            ("12:34:56.789", 1),
            ("10:20:30.189", 2),
            ("2 27:54:32.828", 3),
            ("2 33:44:55.666777", 4),
            ("112233.445566", 5),
            ("1 23:12.1234567", 6),
        ];

        b.iter(|| {
            let cases = test::black_box(&cases);
            for &(s, fsp) in cases {
                let _ = test::black_box(
                    is_duration(s),
                );
            }
        })
    }
}
```

Reply:
Move the `Regex::new` into `lazy_static` will improve the perf a lot:

```
test codec::mysql::duration::benches::bench_is_durtion                                    ... bench:         267 ns/iter (+/- 69)
```

Please considering:

```rust
    use lazy_static::lazy_static;
    use regex::Regex;

    lazy_static! {
        static ref DURATION_REGEX: Regex = Regex::new(
            r"^\s*[-]?(((\d{1,2}\s+)?0*\d{0,3}(:0*\d{1,2}){0,2})|(\d{1,7}))?(\.\d*)?\s*$"
        )
        .unwrap();
    }

    #[inline]
    fn is_duration(s: &str) -> bool {
        DURATION_REGEX.is_match(s)
    }

```

Reply:
After this change, I think current logic is fine.

One more thing: Can we extract the `std::str::from_utf8(arg0).unwrap()` as a new variable so we don't need to do it again in `false` arm?

Reply:
Okay, done

PTAL